File: julia/luhn/luhn.jl
function luhn(text)
    if contains(text, r"[^ 0-9]")
        return false
    end
    num = reverse([parse(Int64, n) for n in split(replace(text, ' '=>""), "")])
    if length(num) < 2
        return false
    end
    num = num .* (repeat([1, 2], (length(num) + 1) ÷ 2))[1:length(num)]
    sum([n < 10 ? n : n - 9 for n in num]) % 10 == 0
end

File: ruby/luhn/luhn.rb
class Luhn
  def self.valid?(text)
    /[^ 0-9]/.match(text) and return false
    digits = (text.gsub(/\s+/, "").each_char.map {|c| c.to_i }).reverse
    digits.length > 1 or return false
    fixed = digits.each.with_index.map { |d, i| i.odd? ? double_with_limit(d) : d }
    fixed.sum % 10 == 0
  end

  private

  def self.double_with_limit(d)
    d > 4 ? d * 2 - 9 : 2 * d
  end
end
